README.md
# Flask-HAL
Flask Extension to easily add support for REST HATEOAS via the HAL Specification:
https://tools.ietf.org/html/draft-kelly-json-hal-07

## With Kim (ideas)

``` python
@app.route('/foo/:id')
def foo_view(id):
    foo = Foo.query.get(id)
    data = FooSerializer().serialize(foo)

    document = Document(links=Links(foo.hal_link()), data=data)

    return document, 200
```


CHANGELOG.rst
Change Log
----------

1.0.3
~~~~~
- Correctly deserialise embedded documents: #25
- Support python setup.py test syntax: #24

1.0.2
~~~~~
- Multiple links with the same relation will be put into an array

1.0.1
~~~~~
- ``self`` links now take into account ``SERVER_NAME`` configuration

1.0.0
~~~~~
- A list can be used in data for Embedded object

2015.10.8
~~~~~~~~~
- Initial release including the core feature set


tests/test_init.py
#!/usr/bin/env python
# encoding: utf-8

"""
tests.test_init
===============

Tests for the init module.
"""

# Standard Libs
import json

# Third Party Libs
from flask import Flask, Response

# First Party Libs
from flask_hal import HAL, HALResponse, document


class TestHAL(object):

    def test_init(self):
        app = Flask(__name__)
        HAL(app)

        assert app.response_class == HALResponse

    def test_can_set_custom_response_class(self):
        app = Flask(__name__)
        HAL(app, Response)

        assert app.response_class == Response


class TestHalResponse(object):

    def test_returns_document_with_hal_document(self):
        app = Flask(__name__)
        with app.test_request_context():
            d = document.Document()
            r = HALResponse.force_type(d, {})

        expected = json.dumps({
            '_links': {
                'self': {
                    'href': '/'
                }
            }
        })

        assert isinstance(r, Response)
        assert r.headers['Content-Type'] == 'application/hal+json'
        assert r.data.decode("utf-8") == expected

    def test_returns_standard_response(self):
        r = HALResponse.force_type(Response('foo'), {})

        assert isinstance(r, Response)
        assert r.headers['Content-Type'] == 'text/html; charset=utf-8'
        assert r.data.decode("utf-8") == 'foo'


tests/test_link.py
#!/usr/bin/env python
# encoding: utf-8

"""
tests.test_link
===============

Unittests for the :module:`flask_hal.link` module.
"""

# Standard Libs
import json

# Third Party Libs
import pytest
from flask import Flask

# First Party Libs
from flask_hal.link import Collection, Link, Self


class TestCollection(object):

    def test_init_raises_type_error(self):
        with pytest.raises(TypeError) as excinfo:
            Collection(
                Link('foo', '/foo'),
                'Foo',
                Link('bar', '/bar'))

        assert 'Foo is not a valid flask_hal.link.Link instance' in str(excinfo.value)

    def test_to_dict(self):
        c = Collection(
            Link('foo', '/foo'),
            Link('bar', '/bar'))

        expected = {
            '_links': {
                'foo': {
                    'href': '/foo'
                },
                'bar': {
                    'href': '/bar'
                }
            }
        }

        assert c.to_dict() == expected

    def test_to_dict_repeat_relation(self):
        c = Collection(
            Link('foo', '/foo'),
            Link('foo', '/bar'))

        expected = {
            '_links': {
                'foo': [
                    {
                        'href': '/foo'
                    },
                    {
                        'href': '/bar'
                    }
                ]
            }
        }

        assert c.to_dict() == expected

    def test_to_json(self):
        c = Collection(
            Link('foo', '/foo'),
            Link('bar', '/bar'))

        expected = json.dumps({
            '_links': {
                'foo': {
                    'href': '/foo'
                },
                'bar': {
                    'href': '/bar'
                }
            }
        })

        assert c.to_json() == expected

    def test_to_json_repeat_relation(self):
        c = Collection(
            Link('foo', '/foo'),
            Link('foo', '/bar'))

        expected = json.dumps({
            '_links': {
                'foo': [
                    {
                        'href': '/foo'
                    },
                    {
                        'href': '/bar'
                    }
                ]
            }
        })

        assert c.to_json() == expected


class TestLink(object):

    def test_only_valid_link_attrs_set(self):
        l = Link('foo', '/foo', foo='foo', name='foo')

        assert not hasattr(l, 'foo')
        assert l.name == 'foo'

    def test_to_dict(self):
        l = Link('foo', '/foo', foo='foo', name='foo')

        expected = {
            'foo': {
                'href': '/foo',
                'name': 'foo',
            }
        }

        assert l.to_dict() == expected

    def test_to_json(self):
        l = Link('foo', '/foo', foo='foo', name='foo')

        expected = json.dumps({
            'foo': {
                'href': '/foo',
                'name': 'foo',
            }
        })

        assert l.to_json() == expected


class TestSelf(object):

    def setup(self):
        self.app = Flask(__name__)

    def test_only_valid_link_attrs_set(self):
        with self.app.test_request_context():
            l = Self(foo='foo', name='foo')

            assert not hasattr(l, 'foo')
            assert l.name == 'foo'

    def test_to_dict(self):
        with self.app.test_request_context():
            l = Self(foo='foo', name='foo')

            expected = {
                'self': {
                    'href': '/',
                    'name': 'foo',
                }
            }

            assert l.to_dict() == expected

    def test_to_json(self):
        with self.app.test_request_context():
            l = Self(foo='foo', name='foo')

            expected = json.dumps({
                'self': {
                    'href': '/',
                    'name': 'foo',
                }
            })

            assert l.to_json() == expected

    def test_with_server_name(self):
        self.app.config['SERVER_NAME'] = 'foo.com'
        with self.app.test_request_context():
            l = Self(foo='foo', name='foo')

            expected = {
                'self': {
                    'href': 'http://foo.com/',
                    'name': 'foo',
                }
            }

            assert l.to_dict() == expected


tests/test_document.py
# Third Party Libs
import flask
import pytest

# First Party Libs
from flask_hal import link
from flask_hal.document import Document, Embedded


def test_document_should_have_link_self():
    app = flask.Flask(__name__)
    with app.test_request_context('/entity/231'):
        document = Document()
        assert flask.request.path == document.links[0].href


def test_should_raise_exception_when_links_are_not_in_collection():
    app = flask.Flask(__name__)
    with app.test_request_context('/entity/231'):
        with pytest.raises(TypeError):
            Document(links=link.Link('foo', 'www.foo.com'))


def test_should_append_embedded_document():
    app = flask.Flask(__name__)
    with app.test_request_context('/entity/231'):
        document = Document(
            embedded={
                'orders': Embedded(
                    embedded={'details': Embedded(
                        data={'details': {}}
                    )},
                    links=link.Collection(
                        link.Link('foo', 'www.foo.com'),
                        link.Link('boo', 'www.boo.com')
                    ),
                    data={'total': 30},
                )
            },
            data={'currentlyProcessing': 14}
        )
        expected = {
            '_links': {
                'self': {
                    'href': flask.request.path
                }
            },
            '_embedded': {
                'orders': {
                    '_links': {
                        'foo': {'href': 'www.foo.com'},
                        'boo': {'href': 'www.boo.com'}
                    },
                    '_embedded': {
                        'details': {'details': {}}
                    },
                    'total': 30,
                }
            },
            'currentlyProcessing': 14
        }
        assert expected == document.to_dict()


def test_document_link_as_list_is_converted_to_collection():
    app = flask.Flask(__name__)
    with app.test_request_context('/entity/231'):
        document = Document(links=[link.Link('foo', 'www.foo.com')])
        assert isinstance(document.links, link.Collection)
        assert 2 == len(document.links)


def test_empty_document_to_json():
    app = flask.Flask(__name__)
    with app.test_request_context('/foo/23'):
        document = Document()
        expected = '{"_links": {"self": {"href": "/foo/23"}}}'
        assert expected == document.to_json()


def test_should_raise_exception_when_embedded_is_not_dict():
    app = flask.Flask(__name__)
    with app.test_request_context('/entity/231'):
        with pytest.raises(TypeError):
            Document(embedded=['details'])


def test_data_in_embedded_can_be_array():
    app = flask.Flask(__name__)
    with app.test_request_context('/entity/231'):
        document = Document(
            embedded={
                'order': Embedded(
                    data=[
                        {
                            'total': 30.00,
                            'currency': 'USD',
                            'status': 'shipped'
                        }, {
                            'total': 20.00,
                            'currency': 'USD',
                            'status': 'processing'
                        }
                    ]
                )
            },
            data={
                'currentlyProcessing': 14
            }
        )
        expected = {
            'currentlyProcessing': 14,
            '_links': {'self': {'href': u'/entity/231'}},
            '_embedded': {
                'order': [{
                    'total': 30.00,
                    'currency': 'USD',
                    'status': 'shipped'
                }, {
                    'total': 20.00,
                    'currency': 'USD',
                    'status': 'processing'
                }]
            }
        }
        assert expected == document.to_dict()


def test_data_in_embedded_can_be_array_of_documents():
    app = flask.Flask(__name__)
    with app.test_request_context('/entity/231'):
        document = Document(
            embedded={
                'order': Embedded(
                    data=[
                        Embedded(
                            data={
                                'total': 30.00,
                                'currency': 'USD',
                                'status': 'shipped'
                            },
                            links=link.Collection(
                                link.Link('foo', 'www.foo30.com'),
                                link.Link('boo', 'www.boo30.com')
                            ),
                        ),
                        Embedded(
                            data={
                                'total': 20.00,
                                'currency': 'USD',
                                'status': 'processing'
                            },
                            links=link.Collection(
                                link.Link('foo', 'www.foo20.com'),
                                link.Link('boo', 'www.boo20.com')
                            ),
                        )
                    ]
                )
            },
            data={
                'currentlyProcessing': 14
            }
        )
        expected = {
            'currentlyProcessing': 14,
            '_links': {'self': {'href': u'/entity/231'}},
            '_embedded': {
                'order': [
                    {
                        '_links': {
                            'foo': {'href': 'www.foo30.com'},
                            'boo': {'href': 'www.boo30.com'}
                        },
                        'total': 30.00,
                        'currency': 'USD',
                        'status': 'shipped'
                    },
                    {
                        '_links': {
                            'foo': {'href': 'www.foo20.com'},
                            'boo': {'href': 'www.boo20.com'}
                        },
                        'total': 20.00,
                        'currency': 'USD',
                        'status': 'processing'
                    }
                ]
            }
        }
        assert expected == document.to_dict()


flask_hal/link.py
#!/usr/bin/env python
# encoding: utf-8

"""
flask_hal.link
==============

Implements the ``HAL`` Link specification.
"""

# Standard Libs
import json

# Third Party Libs
from flask import current_app, request


VALID_LINK_ATTRS = [
    'name',
    'title',
    'type',
    'deprecation',
    'profile',
    'templated',
    'hreflang'
]


class Collection(list):
    """Build a collection of ``HAL`` link objects.

    Example:
        >>> from flask_hal.link import Collection, Link
        >>> l = Collection(
        ...     Link('foo', 'http://foo.com'),
        ...     Link('bar', 'http://bar.com'))
        >>> print l.to_dict()
        ... {
        ...     '_links': {
        ...         'foo': {
        ...             'href": "http://foo.com'
        ...         },
        ...         'bar': {
        ...             'href': 'http://bar.com'
        ...         }
        ...     }
        ... }
    """

    def __init__(self, *args):
        """Initialise a new ``Collection`` object.

        Example:
            >>> l = Collection(
            ...     Link('foo', 'http://foo.com'),
            ...     Link('bar', 'http://bar.com'))

        Raises:
            TypeError: If a link is not a ``flask_hal.link.Link`` instance
        """

        for link in args:
            if not isinstance(link, Link):
                raise TypeError(
                    '{0} is not a valid flask_hal.link.Link instance'.format(link))

            self.append(link)

    def to_dict(self):
        """Returns the Python ``dict`` representation of the ``Collection``
        instance.

        Example:
            >>> from flask_hal.link import Collection, Link
            >>> l = Collection(
            ...     Link('foo', 'http://foo.com'),
            ...     Link('bar', 'http://bar.com'))
            >>> l.to_dict()
            ... {'_links': {'bar': {'href': 'http://bar.com'},
            ... 'foo': {'href': 'http://foo.com'}}}

        Returns:
            dict
        """

        links = {}

        for link in self:
            if link.rel in links.keys():
                if isinstance(links[link.rel], dict):
                    links[link.rel] = [links[link.rel]]
                links[link.rel].append(link.to_dict()[link.rel])
            else:
                links.update(link.to_dict())

        return {
            '_links': links
        }

    def to_json(self):
        """Returns the ``JSON`` representation of the instance.

        Example:
            >>> from flask_hal.link import Collection, Link
            >>> l = Collection(
            ...     Link('foo', 'http://foo.com'),
            ...     Link('bar', 'http://bar.com'))
            >>> l.to_json()
            ... '{"_links":
                    {
                        "foo": {"href": "http://foo.com"},
                        "bar": {"href": "http://bar.com"}
                    }
                }'

        Returns:
            str: The ``JSON`` representation of the instance
        """

        return json.dumps(self.to_dict())


class Link(object):
    """Build ``HAL`` specification ``_links`` object.

    Example:
        >>> from flask_hal.link import Link
        >>> l = Link('foo', 'http://foo.com/bar')
        >>> print l.to_json()
        ... '{"foo": {"href": "http://foo.com/bar"}}'
        >>> l.title = 'Foo'
        >>> print l.to_json()
        ... '{"foo": {"href": "http://foo.com/bar", "name": "Foo"}}'

    """

    def __init__(self, rel, href, **kwargs):
        """Initialise a new ``Link`` object.

        Args:
            rel (str): The links ``rel`` or name
            href (str): The URI to the resource

        Keyword Args:
            name (str): The links name attribute, optional
            title (str): The links title attribute, optional
            type (str): The links type attribute, optional
            deprecation (str): The deprecation attribute, optional
            profile (str): The profile  attribute, optional
            templated (bool): The templated attribute, optional
            hreflang (str): The hreflang attribute, optional
        """

        self.rel = rel
        self.href = href

        for attr in VALID_LINK_ATTRS:
            if attr in kwargs:
                setattr(self, attr, kwargs.pop(attr))

    def to_dict(self):
        """Returns the Python ``dict`` representation of the ``Link`` instance.

        Example:
            >>> from flask_hal.link import Link
            >>> l = Link('foo', 'http://foo.com')
            >>> l.to_dict()
            ... {'foo': {'href': 'http://foo.com'}}

        Returns:
            dict
        """

        # Minimum viable link
        link = {
            'href': self.href
        }

        # Add extra attributes if they exist
        for attr in VALID_LINK_ATTRS:
            if hasattr(self, attr):
                link[attr] = getattr(self, attr)

        return {
            self.rel: link
        }

    def to_json(self):
        """Returns the ``JSON`` encoded representation of the ``Link`` object.

        Example:
            >>> from flask_hal.link import Link
            >>> l = Link('foo', 'http://foo.com', name='Foo')
            >>> print l.to_json()
            ... '{"foo": {"href": "http://foo.com", "name": "Foo"}}'

        Returns:
            str: The ``JSON`` encoded object
        """
        return json.dumps(self.to_dict())


class Self(Link):
    """A class to create the required ``self`` link  from the current
    request URL.
    """

    def __init__(self, **kwargs):
        """Initialises a new ``Self`` link instance. Accepts the same
        Keyword Arguments as :class:`.Link`.

        See Also:
            :class:`.Link`
        """

        url = request.url
        if current_app.config['SERVER_NAME'] is None:
            url = request.url.replace(request.host_url, '/')

        return super(Self, self).__init__('self', url, **kwargs)


flask_hal/__init__.py
#!/usr/bin/env python
# encoding: utf-8

"""
flask_hal
=========

Flask-HAL provides support for the ``HAL`` Specification within Flask allowing
you to build ``REST`` API responses which fulfil this specification.

Read more at the `Official Draft <https://tools.ietf.org/html/draft-kelly-json-hal-07>`_
"""

# Third Party Libs
from flask import Response

# First Party Libs
from flask_hal.document import Document


# Version
__version__ = '0.0.1'


class HAL(object):
    """Enables Flask-HAL integration into Flask Applications, either by the
    Application Factory Pattern or directly into an already created Flask
    Application instance.

    This will set a custom ``response_class`` for the Application which
    handles the conversion of a ``HAL`` document response from a
    view into it's ``JSON`` representation.
    """

    def __init__(self, app=None, response_class=None):
        """Initialise Flask-HAL with a Flask Application. Acts as a proxy
        to :meth:`flask_hal.HAL.init_app`.

        Example:
            >>> from flask import Flask
            >>> from flask_hal import HAL
            >>> app = Flask(__name__)
            >>> HAL(app=app)

        Keyword Args:
            app (flask.app.Flask): Optional Flask application instance
            response_class (class): Optional custom ``response_class``
        """

        if app is not None:
            self.init_app(app, response_class=response_class)

    def init_app(self, app, response_class=None):
        """Initialise Flask-HAL with a Flask Application. This is designed to
        be used with the Flask Application Factory Pattern.

        Example:
            >>> from flask import Flask
            >>> from flask_hal import HAL
            >>> app = Flask(__name__)
            >>> HAL().init_app(app)

        Args:
            app (flask.app.Flask): Flask application instance

        Keyword Args:
            response_class (class): Optional custom ``response_class``
        """

        # Set the response class
        if response_class is None:
            app.response_class = HALResponse
        else:
            app.response_class = response_class


class HALResponse(Response):
    """A custom response class which overrides the default Response class
    wrapper.

    Example:
        >>> from flask import Flask()
        >>> from flask_hal import HALResponse
        >>> app = Flask(__name__)
        >>> app.response_class = HALResponse
    """

    @staticmethod
    def force_type(rv, env):
        """Called by ``flask.make_response`` when a view returns a none byte,
        string or unicode value. This method takes the views return value
        and converts into a standard `Response`.

        Args:
            rv (flask_hal.document.Document): View return value
            env (dict): Request environment

        Returns:
            flask.wrappers.Response: A standard Flask response
        """

        if isinstance(rv, Document):
            return Response(
                rv.to_json(),
                headers={
                    'Content-Type': 'application/hal+json'
                })

        return Response.force_type(rv, env)


flask_hal/document.py
#!/usr/bin/env python
# encoding: utf-8

"""
flask_hal.document
==================

Module for constructing ``HAL`` documents.

Example:
    >>> from flask_hal.document import Document
    >>> d = Document()
    >>> d.to_dict()
"""

# Standard Libs
import json

# First Party Libs
from flask_hal import link


class BaseDocument(object):
    """Constructs a ``HAL`` document.
    """

    def __init__(self, data=None, links=None, embedded=None):
        """Base ``HAL`` Document. If no arguments are provided a minimal viable
        ``HAL`` Document is created.

        Keyword Args:
            data (dict): Data for the document
            links (flask_hal.link.Collection): A collection of ``HAL`` links
            embedded: TBC

        Raises:
            TypeError: If ``links`` is not a :class:`flask_hal.link.Collection`
        """

        self.data = data
        self.embedded = embedded or {}
        self.links = links or link.Collection()

    @property
    def links(self):
        return self._links

    @links.setter
    def links(self, value):
        if not isinstance(value, link.Collection):
            if isinstance(value, (list, set, tuple)):
                value = link.Collection(*value)
            else:
                raise TypeError('links must be a {0} or {1} instance'.format(
                                link.Collection, list))
        self._links = value

    @property
    def embedded(self):
        return self._embedded

    @embedded.setter
    def embedded(self, value):
        if not isinstance(value, dict):
            raise TypeError('embedded must be a {0} instance'.format(dict))
        self._embedded = value

    def to_dict(self):
        """Converts the ``Document`` instance into an appropriate data
        structure for HAL formatted documents.

        Returns:
            dict: The ``HAL`` document data structure
        """

        document = {}

        # Add Data to the Document
        if isinstance(self.data, dict):
            document.update(self.data)

        # Add Links
        if self.links:
            document.update(self.links.to_dict())

        # Add Embedded: Embedded API TBC
        if self.embedded:
            document.update({
                '_embedded': dict(
                    (n, v.to_dict()) for n, v in self.embedded.items()
                )
            })

        return document

    def to_json(self):
        """Converts :class:`.Document` to a ``JSON`` data structure.

        Returns:
            str: ``JSON`` document
        """

        return json.dumps(self.to_dict())


class Document(BaseDocument):
    """Constructs a ``HAL`` document.
    """

    def __init__(self, data=None, links=None, embedded=None):
        """Initialises a new ``HAL`` Document instance. If no arguments are
        provided a minimal viable ``HAL`` Document is created.

        Keyword Args:
            data (dict): Data for the document
            links (flask_hal.link.Collection): A collection of ``HAL`` links
            embedded: TBC

        Raises:
            TypeError: If ``links`` is not a :class:`flask_hal.link.Collection`
        """
        super(Document, self).__init__(data, links, embedded)
        self.links.append(link.Self())


class Embedded(BaseDocument):
    """Constructs a ``HAL`` embedded.

    Example:
        >>> document = Document(
        >>>     embedded={
        >>>         'orders': Embedded(
        >>>             embedded={'details': Embedded(
        >>>                 data={'details': {}}
        >>>             )},
        >>>             links=link.Collection(
        >>>                 link.Link('foo', 'www.foo.com'),
        >>>                 link.Link('boo', 'www.boo.com')
        >>>             ),
        >>>             data={'total': 30},
        >>>         )
        >>>     },
        >>>     data={'currentlyProcessing': 14}
        >>> )
        >>> document.to_json()
        ... {
                "_links": {
                    "self": {"href": "/entity/231"}
                },
                "_embedded": {
                    "orders": {
                        "_embedded": {
                            "details": {"details": {}}
                        },
                        "total": 30,
                        "_links": {
                            "foo": {"href": "www.foo.com"},
                            "boo": {"href": "www.boo.com"}
                        }
                    }
                },
                "currentlyProcessing": 14
            }
    """

    def to_dict(self):
        """Converts the ``Document`` instance into an appropriate data
        structure for HAL formatted documents.

        Returns:
            dict: The ``HAL`` document data structure
        """
        if isinstance(self.data, (list, tuple, set)):
            data = []

            for item in self.data:
                if isinstance(item, BaseDocument):
                    data.append(item.to_dict())
                else:
                    data.append(item)
            return data

        return super(Embedded, self).to_dict()
